src/dnd/mod.rs

```diff
@@ -0,0 +1,278 @@
+}
+
+#[derive(Debug)]
+pub enum OfferEvent<T> {
+    Enter {
+        x: f64,
+        y: f64,
+        mime_types: Vec<MimeType>,
+        surface: T,
+    },
+    Motion {
+        x: f64,
+        y: f64,
+    },
+    /// The offer is no longer on a DnD destination.
+    LeaveDestination,
+    /// The offer has left the surface.
+    Leave,
+    /// An offer was dropped
+    Drop,
+    /// If the selected action is ASK, the user must be presented with a choice.
+    /// [`Clipboard::set_action`] should then be called before data can be
+    /// requested and th DnD operation can be finished.
+    SelectedAction(DndAction),
+    Data {
+        data: Vec<u8>,
+        mime_type: MimeType,
+    },
+}
+
+/// A rectangle with a logical location and size relative to a [`DndSurface`]
+#[derive(Debug, Default, Clone)]
+pub struct Rectangle {
+    pub x: f64,
+    pub y: f64,
+    pub width: f64,
+    pub height: f64,
+}
+
+impl Rectangle {
+    fn contains(&self, x: f64, y: f64) -> bool {
+        self.x <= x && self.x + self.width >= x && self.y <= y && self.y + self.height >= y
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct DndDestinationRectangle {
+    /// A unique ID
+    pub id: u128,
+    /// The rectangle representing this destination.
+    pub rectangle: Rectangle,
+    /// Accepted mime types in this rectangle
+    pub mime_types: Vec<MimeType>,
+    /// Accepted actions in this rectangle
+    pub actions: DndAction,
+    /// Prefered action in this rectangle
+    pub preferred: DndAction,
+}
+
+pub enum DndRequest<T> {
+    /// Init DnD
+    InitDnd(Box<dyn crate::dnd::Sender<T> + Send>),
+    /// Register a surface for receiving Dnd events.
+    Surface(DndSurface<T>, Vec<DndDestinationRectangle>),
+    /// Start a Dnd operation with the given source surface and data.
+    StartDnd {
+        internal: bool,
+        source: DndSurface<T>,
+        icon: Option<Icon<DndSurface<T>>>,
+        content: Box<dyn AsMimeTypes + Send>,
+        actions: DndAction,
+    },
+    /// Peek the data of an active DnD offer
+    Peek(MimeType),
+    /// Set the DnD action chosen by the user.
+    SetAction(DndAction),
+    /// End an active DnD Source
+    DndEnd,
+}
+
+#[derive(Debug)]
+pub enum DndEvent<T> {
+    /// Dnd Offer event with the corresponding destination rectangle ID.
+    Offer(Option<u128>, OfferEvent<T>),
+    /// Dnd Source event.
+    Source(SourceEvent),
+}
+
+impl<T> Sender<T> for calloop::channel::Sender<DndEvent<T>> {
+    fn send(&self, t: DndEvent<T>) -> Result<(), SendError<DndEvent<T>>> {
+        self.send(t)
+    }
+}
+
+impl<T> Sender<T> for calloop::channel::SyncSender<DndEvent<T>> {
+    fn send(&self, t: DndEvent<T>) -> Result<(), SendError<DndEvent<T>>> {
+        self.send(t)
+    }
+}
+
+pub enum Icon<S> {
+    Surface(S),
+    /// Argb8888 or Xrgb8888 encoded image data pre-multiplied by alpha.
+    Buf {
+        width: u32,
+        height: u32,
+        data: Vec<u8>,
+        transparent: bool,
+    },
+}
```

Comment:
To properly support hiDPI, this also needs a scale factor, right?

A `scale: f64` field would make sense, though I guess that wouldn't perfectly represent a `wp_fractional_scale_v1::preferred_scale` so it could cause rounding errors with fractional scales.